skfem/mesh/mesh_tet_1.py

```diff
@@ -34,18 +35,36 @@ def element_finder(self, mapping=None):
         if mapping is None:
             mapping = self._mapping()
 
-        tree = cKDTree(np.mean(self.p[:, self.t], axis=1).T)
+        if not hasattr(self, '_cached_tree'):
+            self._cached_tree = cKDTree(np.mean(self.p[:, self.t], axis=1).T)
+
+        tree = self._cached_tree
+        nelems = self.t.shape[1]
+
+        def finder(x, y, z, _exhaustive=False):
+
+            if _exhaustive or x.shape[0] > nelems:
+                # search all elements if forced by the flag or if there is a
+                # large number of points
+                ix = None
+            else:
+                ix = tree.query(np.array([x, y, z]).T,
+                                min(5, nelems))[1].flatten()
 
-        def finder(x, y, z):
-            ix = tree.query(np.array([x, y, z]).T, 5)[1].flatten()
             X = mapping.invF(np.array([x, y, z])[:, None], ix)
-            inside = (
-                (X[0] >= 0) *
-                (X[1] >= 0) *
-                (X[2] >= 0) *
-                (1 - X[0] - X[1] - X[2] >= 0)
-            )
-            return np.array([ix[np.argmax(inside, axis=0)]]).flatten()
+            inside = ((X[0] >= 0) *
+                      (X[1] >= 0) *
+                      (X[2] >= 0) *
+                      (1 - X[0] - X[1] - X[2] >= 0))
+            elems = np.argmax(inside, axis=0)
+
+            if ix is not None and not inside[elems].any(axis=0).all():
+                warnings.warn("Falling back to exhaustive search because the "
```

Comment:
I verified that this is activated only in `test_finder_simplex` where some of the elements have a nasty shape.